src/Generator/EnumGenerator/Cases/CaseFactory.php

```diff
@@ -0,0 +1,33 @@
+<?php
+
+namespace Laminas\Code\Generator\EnumGenerator\Cases;
+
+use function array_key_exists;
+use function assert;
+
+/** @internal */
+final class CaseFactory
+{
+    /**
+     * @psalm-param array{
+     *      name: non-empty-string,
+     *      pureCases: list<non-empty-string>,
+     * }|array{
+     *      name: non-empty-string,
+     *      backedCases: array{
+     *          type: 'int'|'string',
+     *          cases: array<non-empty-string, int|non-empty-string>,
+     *      },
+     * } $options
+     * @return BackedCases|PureCases
+     */
+    public static function fromOptions(array $options)
```

Comment:
Personally I try to avoid abstract classes whenever possible, and for this particular case, I think putting the "factory" as part of the abstract class is not very extensible.

Next year, when PHP 7.4 is dropped (assuming that's the plan) we can use Union types instead.